src/auth/auth.service.ts

```diff
@@ -0,0 +1,102 @@
+import { ForbiddenException, Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { JwtService } from '@nestjs/jwt';
+import { PrismaService } from 'src/database/prisma.service';
+import { CreateUserDto } from './input/auth.dto';
+import { Tokens } from './models/tokens.type';
+import * as bcrypt from 'bcrypt';
+import { JwtPayload } from './models/jwt-payload.type';
+import { UsersService } from '../users/users.service';
+import { DatabaseRecord } from '../common/models/databaseRecord.model';
+import { User } from '../users/models/user.type';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private prisma: PrismaService,
+    private jwtService: JwtService,
+    private config: ConfigService,
+    private usersService: UsersService,
+  ) {}
+
+  async register(dto: CreateUserDto): Promise<Tokens> {
+    const user: DatabaseRecord<User> = await this.usersService.createUser(dto);
+
+    const tokens = await this.getTokens(user.id, user.email);
+    await this.updateRtHash(user.id, tokens.refresh_token);
+
+    return tokens;
+  }
+
+  async login(dto: CreateUserDto): Promise<Tokens> {
+    const user = await this.usersService.getUserByEmail(dto.email);
+
+    if (!user) throw new ForbiddenException('Access Denied');
+
+    const passwordMatches = await bcrypt.compare(dto.password, user.hash);
+    if (!passwordMatches) throw new ForbiddenException('Access Denied');
+
+    const tokens = await this.getTokens(user.id, user.email);
+    await this.updateRtHash(user.id, tokens.refresh_token);
+
+    return tokens;
+  }
+
+  async logout(userId: string): Promise<boolean> {
+    await this.prisma.user.updateMany({
+      where: {
+        id: userId,
+        hashedRt: {
+          not: null,
+        },
+      },
+      data: {
+        hashedRt: null,
+      },
+    });
+    return true;
+  }
+
+  async refreshTokens(userId: string, rt: string): Promise<Tokens> {
+    const user = await this.usersService.getUserById(userId);
+    if (!user || !user.hashedRt) throw new ForbiddenException('Access Denied');
+
+    const rtMatches = await bcrypt.compare(user.hashedRt, rt);
+    if (!rtMatches) throw new ForbiddenException('Access Denied');
+
+    const tokens = await this.getTokens(user.id, user.email);
+    await this.updateRtHash(user.id, tokens.refresh_token);
+
+    return tokens;
+  }
+
+  async updateRtHash(userId: string, rt: string): Promise<void> {
+    const salt = await bcrypt.genSalt();
+    const hash = await bcrypt.hash(rt, salt);
+
+    await this.usersService.updateUser<string>(userId, 'hashedRt', hash);
+  }
+
+  async getTokens(userId: string, email: string): Promise<Tokens> {
+    const jwtPayload: JwtPayload = {
+      sub: userId as any,
```

Comment:
Avoid casts to `any`. Now that I look at it - instead of `userId as any`, how about setting `sub: string` in `JwtPayload`? Generally `sub` is a string and you have a `number`. `sub` can be anything uniquely identifying the principal. Example would be a user email address.